### DISCUSSION TITLE
404 Page not found error on page reload or refresh in production mode

### DISCUSSION THREAD

**Discussion by PatrickNaenae** [vercel/next.js#51646] [Help] [answered]
### Summary

I get a 404 error when i reload or refresh my pages in production. I guess it has to be a server error but i don't know the solution to it.

![Screen Shot 2023-06-22 at 12 00 37](https://github.com/vercel/next.js/assets/83951220/c60793e2-ab89-4ee3-91c9-f3bd55b6de68)

**Comment by icyJoseph** [comment]
Hi,

Is your app a Static Export? 

I wonder if the server is looking for `/main/services.html`, but you are requesting `/main/services` - if you add `.html` to the end it will work. https://mozfin.com/main/services.html

So you gotta route content so that it can pick up on the `.html` extension - alternatively you can use trailing slashes, but wait with that, let me find some data for you.

**Reply by icyJoseph** [reply]
- https://nextjs.org/docs/app/building-your-application/deploying/static-exports#deploying

For example they do:

```
server {
  listen 80;
  server_name acme.com;
 
  root /var/www;
 
  location / {
      try_files /out/index.html =404;
  }
 
  location /blog/ {
      rewrite ^/blog/(.*)$ /out/blog/$1.html break;
  }
 
  error_page 404 /out/404.html;
  location = /404.html {
      internal;
  }
}
```

And that'll serve `/blob/post-1.html` to `/blog/post-1` routes - you need to figure out how to do that in your server.

**Reply by PatrickNaenae** [reply]
Hi Joseph,

Thank you for your reply. That means i will need to create a server.js file in my root directory and add that?

**Reply by icyJoseph** [reply]
Mmm no, most static asset server solutions can handle this. You are using Next.js in  static export mode, right?

**Reply by icyJoseph** [reply]
Ah you are using `Microsoft-IIS/10.0`, not sure how to do this there. I think there's a like a web.config file or something?

**Reply by PatrickNaenae** [reply]
Yes I am using Next.js in static export mode. So what should i do?. I added a .html extension manually to the url like you did above "https://mozfin.com/main/services.html" and a refresh didn't cause the 404 error.

**Reply by PatrickNaenae** [reply]
No, I am using smarterASP.NET

**Reply by icyJoseph** [reply]
The "problem" is that the static export mode, without trailingSlashes, outputs, for example, `/main/service.html`, so a request for exactly that, has to arrive, otherwise your server rejects it.

We need to tell the server to serve `/main/service.html`, when `/main/service` is requested. 

> With trailing slashes next outputs `/main/services/index.html`, but this is not a real fix, it is a hack (unless you did want trailing slashes) and you'd need to update all of your links, might even hurt SEO (don't quote me on that)

After you have fixed this, you also need to route 404s to the 404.html page, that's at the root of your `out` folder.

**Reply by PatrickNaenae** [reply]
Wow. So if trailingSlashes is not a real fix, what then is the best fix?

**Reply by icyJoseph** [reply]
To fix your static server!

**Reply by icyJoseph** [reply]
Do you have any server configurations? Have you migrated from CRA/Vite to Next.js? Maybe you have a faulty config file in your server.

**Reply by PatrickNaenae** [reply]
I don't even have any server configuration atall.

**Reply by icyJoseph** [reply]
Looks like smarterASP.NET supports an `.htaccess` file.

- https://www.smarterasp.net/support/kb/a1964/how-to-set-up-rules-and-redirects-in-_htaccess.aspx?KBSearchID=1109780

```.htaccess
RewriteEngine On
RewriteCond %{REQUEST_FILENAME} !-f
RewriteRule ^([^\.]+)$ $1.html [NC,L]
```

Here's a longer form: https://gist.github.com/davidvandenbor/f5a2c18c472ceb68d0dd

**Reply by PatrickNaenae** [reply]
Thank you for your replies Joesph. Let me try out the configuration. I will let you know how it goes.

**Reply by icyJoseph** [reply]
Awesome! Otherwise, try to reach out for customer support on that site. They should be able to guide you. Just to sum up the problem is the Static Host Server configuration. As of now it wants exact matches for the requests, but we want `/main/service` to show the contents of `/main/service.html`.

**Reply by PatrickNaenae** [reply]
Hi Joseph, Thank you so much for your assistance. So I reached out to their customer support team and they fixed it for me

**Reply by icyJoseph** [reply]
Awesome! Please consider marking this as answered. Have a good day!

**Reply by PatrickNaenae** [reply, marked as answer]
Ok. I will do that. Thank you
